pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -0,0 +1,215 @@
+    args = node.args
+    annotations = (
+        args.posonlyargs_annotations + args.annotations + args.kwonlyargs_annotations
+    )
+    if args.vararg is not None:
+        annotations.append(args.varargannotation)
+    if args.kwarg is not None:
+        annotations.append(args.kwargannotation)
+    return annotations
+
+
+def _has_valid_annotations(
+    annotations: list[astroid.NodeNG | None],
+) -> list[astroid.NodeNG | None]:
+    for annotation in annotations:
+        if annotation is not None:
+            return True
+    return False
+
+
+class HassTypeHintChecker(BaseChecker):  # type: ignore[misc]
+    """Checker for setup type hints."""
+
+    __implements__ = IAstroidChecker
+
+    name = "hass_enforce_type_hints"
+    priority = -1
+    msgs = {
+        "W0020": (
+            "Argument %d should be of type %s",
+            "hass-argument-type",
+            "Used when method argument type is incorrect",
+        ),
+        "W0021": (
+            "Return type should be %s",
+            "hass-return-type",
+            "Used when method return type is incorrect",
+        ),
+    }
+    options = ()
+
+    def __init__(self, linter: PyLinter | None = None) -> None:
+        super().__init__(linter)
+        self.current_package: str | None = None
+        self.module: str | None = None
+
+    def visit_module(self, node: astroid.Module) -> None:
+        """Called when a Module node is visited."""
+        self.module = node.name
+        if node.package:
+            self.current_package = node.name
+        else:
+            # Strip name of the current module
+            self.current_package = node.name[: node.name.rfind(".")]
+
+    def visit_functiondef(self, node: astroid.FunctionDef) -> None:
+        """Called when a FunctionDef node is visited."""
+        for match in _METHOD_MATCH:
+            self._visit_functiondef(node, match)
+
+    def visit_asyncfunctiondef(self, node: astroid.AsyncFunctionDef) -> None:
+        """Called when a FunctionDef node is visited."""
+        for match in _METHOD_MATCH:
+            self._visit_functiondef(node, match)
+
+    def _visit_functiondef(
+        self, node: astroid.FunctionDef, match: TypeHintMatch
+    ) -> None:
+        if node.name != match.function_name:
+            return
+        if node.is_method():
+            return
+        if not match.module_filter.match(self.module):
+            return
+
+        # Check that at least one argument is annotated.
+        annotations = _get_all_annotation(node)
+        if node.returns is None and not _has_valid_annotations(annotations):
+            return
+
+        # Check that all arguments are correctly annotated.
+        for key, value in match.arg_types.items():
+            if not value.check(annotations[key]):
+                self.add_message(
+                    "hass-argument-type",
+                    node=node.args.args[key],
+                    args=(key, value.expected_type),
```

Comment:
Let's be a bit more friendly to people.
```suggestion
                    args=(key + 1, value.expected_type),
```

Reply:
Don't we all count arguments from 0 !?!